---
title: "Analyze Protein Categories"
subtitle: ""
author: "[Benjamin Dubreuil](https://github.com/benjamin-elusers/)"
home: "https://github.com/benjamin-elusers/yeastomics/"
github: "benjamin-elusers"
mail: "benjamin.dubreuil@weizmann.ac.il"
date: "`r format(Sys.time(), 'Last Modified: %d-%b-%Y %R %p')`"
output:
  rmdformats::downcute:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: true
    highlight: tango
#  html_document: 
    code_download: true
    code_folding: hide
    fig_caption: yes
    keep_md: yes
#    toc: yes
    toc_depth: 3
    toc_float:
      toc_collapsed: true
    number_sections: true
    theme: lumen
  editor_options: 
    chunk_output_type: console
  bibliography: bibliography.bib
  csl: biomed-central.csl
runtime: shiny

---
<!-- Github Logo -->
<a href="https://github.com/benjamin-elusers/yeastomics/" class="github-corner" aria-label="View source on Github">
  <svg width="80" height="80" viewBox="0 0 250 250" style="fill:#69b3a2; color:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true">
    <path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path>
    <path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path>
    <path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path>
  </svg>
</a>

<style> 
.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}
@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}
@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      collapse = TRUE,
                      cache.lazy = FALSE,
                      autodep = TRUE,
                      warning = FALSE)

suppressPackageStartupMessages({
  library(tictoc)
  library(tidyverse)
  library(stringr)
  library(broom)
  library(hablar)
  library(ggplot2)
  library(ggnewscale)
  library(ggpubr)
  library(ggthemes)
  library(ggsci)
  library(cowplot)
  theme_set(theme_cowplot())

  source("src/utils.r")
  source("src/function_annotation.r")
  source("src/function_sequence.r")
  source("src/function_phylogenetic.r")
  source("src/function_analysis.r")
  source("src/function_datalocal.r")
  source("src/function_datapub.r")
})

# turn off annoying messages from dplyr::summarise
options(dplyr.summarise.inform = FALSE, dplyr.width=Inf)
```


##  Properties & feature

```{r protcat}
# protein properties (qualitative)
PROP=readRDS(get.last.file("output","proteome-properties")) %>% 
          dplyr::select(-contains("chemsig_X"))

# protein features (quantitative)
FEAT = readRDS(get.last.file("output","proteome-features"))

# Matrix format of properties and features
PROP.m = PROP %>% 
      ungroup() %>% 
      dplyr::select( where(is.logical) )  %>% 
     hablar::convert(num(everything())) %>%
      as.matrix

FEAT.m = FEAT %>%
        ungroup() %>% 
        dplyr::select( where(is.numeric) ) %>%
      as.matrix

TOT_PROP=ncol(PROP.m)
TOT_FEAT=ncol(FEAT.m)

prop = PROP %>% 
        pivot_longer(cols = starts_with('cat_'),
                     names_to = c('categories','source','property'), 
                     names_pattern="cat_(.+)\\.(.+)\\.(.+)",
                     values_to = "has_prop") %>%
        mutate(col_prop = paste0(categories,'.',source,'.',property))

feat = FEAT %>% 
        pivot_longer(cols = starts_with('cat_'),
                     names_to = c('categories','source','feature'), 
                     names_pattern="cat_(.+)\\.(.+)\\.(.+)",
                     values_to = "value") %>%
        mutate(col_feat = paste0(categories,'.',source,'.',feature))

```


# Expression data

```{r expdata}

ABUNDANCE = load.ho2018.data() %>% 
            dplyr::select(orf, MPC = mean.mpc, gfp=GFP.avg,ms=MS.avg) %>% # select the average expression
            group_by(orf) %>% dplyr::summarise(across(where(is.numeric),log10)) %>% # apply log10 to expression
            rowwise()  %>% mutate( noval = mean(is.na(c_across(2:4))) ) %>% # check how many expression values
            dplyr::filter(noval<1) %>% dplyr::select(-noval) # remove no values

EXPRESSION = FEAT %>% 
             dplyr::select(ORF,contains(c("tAI","CU_cai","CU_fop","paxdb","TE_exponential"))) %>%
             dplyr::select(-paste0("cat_transcriptomics.paxdb.ortho_ppm_", c("sd","cv","max","min","n")))%>%
             dplyr::rename_with(.cols=everything(), str_replace, pattern="cat_.+\\..+\\.(.+)$",replacement="\\1")

colab = c('PPM','MPC','gfp','ms')
colexp = c("tAI","CU_cai","CU_fop","TE_exponential","TE_exponential_A14201","TE_exponential_gb15")
colortho = c('ppm_4932','ppm_214684','ppm_4896','ppm_5061','ortho_ppm_avg')
exp.type = rep(c('PROTEIN','GENE','ORTHOLOG'), times=c(length(colab),length(colexp),length(colortho)))

```

```{r evodata}
# EVOLUTIONARY RATE ON FUNGI
FUNGI = load.dubreuil2021.data(1) %>% 
     as_tibble() %>%
     dplyr::select(c(starts_with('EVO.'),'PPM','ORF','UNIPROT')) %>%
     hablar::convert(dbl(EVO.FULL), dbl(PPM)) %>% 
     ungroup() %>% 
     left_join(ABUNDANCE,by=c('ORF'='orf'))

# EVOLUTIONARY RATE ON BOTH FUNGI AND STRAINS
STRAINS = readRDS("data/PROTEIN-EVO-FUNGI-SNP.rds") %>%
       dplyr::select(c(starts_with(c('EVO.','SNP.')),'PPM','ORF','UNIPROT','IS_FUNGI','IS_STRAINS')) %>% 
       group_by(ORF) %>% dplyr::mutate(across(starts_with("SNP."),log)) %>% # apply log10 to SNP rate4site
       left_join(ABUNDANCE,by=c('ORF'='orf')) %>% 
       left_join(EXPRESSION,by=c('ORF'))


EVO = STRAINS %>% 
      dplyr::filter(IS_FUNGI & IS_STRAINS) %>% 
      pivot_longer( cols=starts_with(c("EVO.","SNP.")), names_pattern="(.+)\\.(.+)", names_to=c('TIMESCALE','REGION'), values_to='R4S' ) %>% 
      pivot_longer( cols=all_of(c(colab,colexp,colortho)), names_to='EXPRESSION',values_to='ABUNDANCE') %>%
      mutate(FROM = factor(EXPRESSION,levels = c(colab,colexp,colortho), labels = exp.type) )
#
COR = cor.sub.by(EVO,XX = 'ABUNDANCE',YY = 'R4S', BY = c('TIMESCALE','EXPRESSION','FROM','REGION') ) %>% 
    mutate(toshow=sprintf("\n r %.3f \n p %.1e \n N %s \n",r,p,n), 
           y.touse=ifelse(TIMESCALE=='EVO',-Inf,Inf))
```

### F1 : EVO vs EXP correlation

```{r F1-evo-exp}
# FIGURE 1 : EVOLUTION BY TIMESCALE AND BY PROTEIN EXPRESSION MEASURES
COR_fullprot = COR %>% dplyr::filter(REGION=='FULL' & FROM=="PROTEIN")
EVO_fullprot = EVO %>% dplyr::filter(REGION=='FULL' & FROM=="PROTEIN")
P1 = ggplot(EVO_fullprot,aes(x=ABUNDANCE,y=R4S,col=EXPRESSION)) + 
  geom_rect(data=COR_fullprot, aes(fill=EXPRESSION),xmin=-Inf,xmax=Inf,ymin=-Inf,ymax=Inf,alpha=0.3,inherit.aes=F) + scale_fill_aaas() + 
  geom_point(size=0.1,alpha=0.2,color='black') + geom_smooth(method='gam',size=0.5) + scale_color_d3() + 
  new_scale_fill()  +
  stat_density_2d(aes(fill = ..level..), geom = "polygon", colour=NA,size=0.25,alpha=0.1) + scale_fill_distiller(palette = 'Blues',direction = -1)+
  geom_text(COR_fullprot, mapping=aes(label=toshow),
             #label.padding=unit(0.05,'lines'),label.r=unit(0.05,'lines'),label.size=0,
             color="gray30",alpha=1,size=2.9,
             x=Inf, y=Inf, hjust='inward',vjust='inward') +
  facet_grid(TIMESCALE~EXPRESSION,scales = 'free') + theme_pubclean() +
  theme(legend.position = 'none') 
cowplot::save_plot(P1,filename='output/F1.evorate-abundance_by_timescale-prot.exp.png',base_height = 5, base_width = 10)
plot(P1)

# FIGURE S1 : EVOLUTION BY TIMESCALE AND BY GENE EXPRESSION MEASURES
COR_fullgene = COR %>% dplyr::filter(REGION=='FULL' & FROM=="GENE")
EVO_fullgene = EVO %>% dplyr::filter(REGION=='FULL' & FROM=="GENE")
S1 = ggplot(EVO_fullgene,aes(x=ABUNDANCE,y=R4S,col=EXPRESSION)) + 
  geom_rect(data=COR_fullgene, aes(fill=EXPRESSION),xmin=-Inf,xmax=Inf,ymin=-Inf,ymax=Inf,alpha=0.3,inherit.aes=F) + scale_fill_aaas() + 
  geom_point(size=0.1,alpha=0.2,color='black') + geom_smooth(method='gam',size=0.5) + scale_color_d3() + 
  new_scale_fill()  +
  stat_density_2d(aes(fill = ..level..), geom = "polygon", colour=NA,size=0.25,alpha=0.1) + scale_fill_distiller(palette = 'Blues',direction = -1)+
  geom_text(COR_fullgene, mapping=aes(label=toshow),
             #label.padding=unit(0.05,'lines'),label.r=unit(0.05,'lines'),label.size=0,
             color='gray30',alpha=1,size=2.9,
             x=Inf,y=Inf, hjust='inward',vjust='inward') +
  facet_grid(TIMESCALE~EXPRESSION,scales = 'free') + theme_pubclean() +
  theme(legend.position = 'none') 
cowplot::save_plot(S1,filename='output/S1.evorate-abundance_by_timescale-gene.exp.png',base_height = 5, base_width = 10)
plot(S1)
```


# Protein categories

There are a total of **`r TOT_PROP`** proteome properties.
We also gathered **`r TOT_FEAT`** protein features.

### F2 : Non-conserved properties

```{r F2-non_cons-prop }
prop.size = prop %>% 
  mutate(has_ortho = ORF %in% STRAINS$ORF[STRAINS$IS_FUNGI],
         has_snp = ORF %in% STRAINS$ORF[STRAINS$IS_STRAINS]) %>%
  group_by(categories,source,property) %>% 
  summarise( in_prot=sum(has_prop), 
             in_ortho=sum(has_prop*has_ortho),
             in_snp=sum(has_prop*has_snp),
             r_prot_ortho = in_ortho/in_prot,
             r_prot_snp = in_snp/in_prot)

prop_nocons = prop.size %>% filter(in_ortho<10 & in_prot > 12)
# 1. Properties under-represented in yeast orthologs
P2 = ggplot(prop_nocons ) + 
  geom_bar(aes(y=in_prot, x=reorder(property,in_prot),fill=source),stat='identity') +
  geom_bar(aes(y=in_ortho, x=reorder(property,in_prot)),stat='identity') + 
  geom_hline(yintercept=10,color='black',linetype='dashed') +
  facet_wrap(~categories,scales = 'free_y') + ylab('# Proteins') +
  coord_flip() + theme( axis.text = element_text(size=12), legend.position = 'top') 
#save_plot(path="output/", filename = 'F2.properties-absent-ortho.pdf', plot = P2, device = 'pdf', scale = 3 )
save_plot(path="output/", filename = 'F2.properties-absent-ortho.png', plot = P2, device = 'png',base_height = 10, base_width = 20)
plot(P2)

categories=unique(prop_nocons$categories)
for( icat in seq_along(categories) ){
  CAT = categories[icat]
  S2 = ggplot(prop_nocons %>% filter(categories == CAT )) + 
    geom_bar(aes(y=in_prot, x=reorder(property,in_prot),fill=source),stat='identity') +
    geom_bar(aes(y=in_ortho, x=reorder(property,in_prot)),stat='identity') + 
    geom_hline(yintercept=10,color='black',linetype='dashed') +
    facet_wrap(~categories,scales = 'free_y',ncol=2) + ylab('# Proteins') +
    coord_flip() + theme( axis.text = element_text(size=10)) +
    scale_x_discrete(position = "top", name='') + ylim(0,100) + scale_fill_npg()
  save_plot(path="output/", filename = sprintf('S2_%s.properties-absent-ortho-%s.png',icat,CAT), plot = S2,device = 'png', scale = 2 )
 # save_plot(path="output/", filename = sprintf('S2.%s.properties-absent-ortho-%s.pdf',icat,CAT), plot = S2,device = 'pdf', scale = 2 )
  plot(S2)
}
```

# Control for expression 

```{r fit-evo-exp}
# Fitting mean Evolutionary rate (R) to Protein abundance (E)
init.params =  list(Asym = 3, xmid = 1.5, scal=-1)
evo_mpc.fit = STRAINS %>%
               broom::augment_columns(x=nls(EVO.FULL ~ SSlogis(MPC,Asym,xmid,scal),start=init.params,data=.)) %>% 
               mutate(SSR = sum(.resid^2), SR = sum(.resid)) %>%
               dplyr::rename_with(.fn=xxS, sx='evo',.cols=c(".fitted", ".resid", "SSR", "SR") )

snp_mpc.fit = STRAINS %>%
               broom::augment_columns(x=lm(SNP.FULL ~ MPC,data=.)) %>% 
               mutate(SSR = sum(.resid^2), SR = sum(.resid)) %>% 
               dplyr::rename_with(.fn=xxS, sx='snp', .cols=c(".fitted", ".resid", "SSR", "SR") )


propfit = inner_join(evo_mpc.fit,snp_mpc.fit, suffix=c('.evo','.snp')) %>% 
              left_join(prop) %>% dplyr::filter(has_prop)

.res_prop = propfit %>% 
              group_by(col_prop) %>% 
              mutate(  N=n(),
                .res_evo.neg = sum(.resid.evo<0), 
                .res_evo.pos = sum(.resid.evo>0),
                .res_evo.dN  = .res_evo.neg-.res_evo.pos,
                .res_evo.rN  = .res_evo.pos/.res_evo.neg,
                .res_evo.SR  = sum(.resid.evo),
                .res_evo.SR.log = sign(.res_evo.SR) * log(abs(.res_evo.SR)),
                .res_evo.SR.neg = sum(.resid.evo[.resid.evo<0]), 
                .res_evo.SR.pos = sum(.resid.evo[.resid.evo>0]),
                .res_evo.SR.avg = mean_(.resid.evo), 
                
                .res_snp.neg = sum(.resid.snp<0), 
                .res_snp.pos = sum(.resid.snp>0),
                .res_snp.dN  = .res_snp.neg-.res_snp.pos,
                .res_snp.rN  = .res_snp.pos/.res_snp.neg,
                .res_snp.SR  = sum(.resid.snp),
                .res_snp.SR.log = sign(.res_snp.SR) * log(abs(.res_snp.SR)),
                .res_snp.SR.neg = sum(.resid.snp[.resid.snp<0]), 
                .res_snp.SR.pos = sum(.resid.snp[.resid.snp>0]),
                .res_snp.SR.avg = mean_(.resid.snp), 
           ) %>%
  arrange(desc(.res_evo.SR.avg)) %>%
  dplyr::select(categories,source,property, col_prop, N, starts_with('.res_')) %>%
  distinct() %>% 
  group_by(categories) %>% mutate(rSR = dense_rank(.res_evo.SR.avg) , ncat=n())

S3 = ggplot(.res_prop, 
               aes(y = reorder(str_trunc(property,w=50), -.res_evo.SR.avg),
                   x = .res_evo.SR.avg, fill=source)) +
      geom_bar(stat='identity',position='dodge') + 
      theme( axis.text = element_text(size=5))
plot(S3)
save_plot(plot = S3, 'output/S3.evorate-residuals-properties.pdf', base_height=50, base_width=20,limitsize = FALSE)

categories=unique(prop_nocons$categories)
for( icat in seq_along(categories) ){
  CAT = categories[icat]
  S3 = ggplot(.res_prop %>% filter(categories == CAT & N>9 ), 
              aes(y = reorder(str_trunc(property,w=50), -.res_evo.SR.avg), x = .res_evo.SR.avg, fill=source)) +
      geom_bar(stat='identity',position='dodge') + 
      theme( axis.text = element_text(size=5))
  plot(S3)
  save_plot(path="output/", filename = sprintf('S3_%s.properties-residuals-%s.png',icat,CAT), plot = S3,device = 'png', scale = 2 )

}


F3 = ggplot(.res_prop %>% dplyr::filter(N>9) %>% slice_min_max( order_by=.res_evo.SR.avg, n=10 ), 
               aes(y = reorder(str_trunc(property,w=50), -.res_evo.SR.avg),
                   x = .res_evo.SR.avg, fill=source)) +
      geom_bar(stat='identity',position='dodge')  + 
      geom_text(aes(label=N),hjust='inward',size=5,col='black') +
        facet_wrap(~categories,scales = 'free_y') +
        theme( axis.text = element_text(size=12)) + xlab('Mean residual evolutionary rate') + ylab('Properties')
plot(F3)
save_plot(plot = F3, 'output/F3.evorate-residuals-properties-top10.png', base_height=12, base_width=20)

.res_prop %>% 
  dplyr::select( categories, source, property,N ,.res_evo.SR,.res_evo.SR.avg,ncat,rSR) %>%
  dplyr::filter( categories == 'biophysics' & N > 9) %>% arrange(rSR) %>% print(n=250)


feat.rk = left_join(STRAINS,feat) %>%
             dplyr::filter(!is.na(value) & !is.na(SNP.FULL) & !is.na(EVO.FULL) & !is.na(MPC)) %>%
             group_by(col_feat) %>% 
             mutate(feat_n = n(), rk.value = dense_rank(value), rk.mpc = dense_rank(MPC))

semifit.res = inner_join(evo_mpc.fit,snp_mpc.fit, suffix=c('.evo','.snp')) %>% 
              left_join(feat.rk)

test = semifit.res %>% 
  group_by(col_feat) %>% 
  mutate( R2_evo=scor(.resid.evo,value)$estimate,  
          R2_snp=scor(.resid.snp,value)$estimate) %>%
  group_by(categories) %>% mutate( ncat=n(), 
                                   r.R2_evo=dense_rank(R2_evo),
                                   r.R2_snp=dense_rank(R2_snp)) %>% 
  summarise(categories,source,feature,feat_n,R2_evo,R2_snp, r.R2_evo,r.R2_snp,ncat) %>% distinct()

F4 = ggplot(test %>% dplyr::filter(feat_n>100),# %>% slice_min_max( order_by=.res_evo.SR.avg, n=10 ), 
               aes(y = reorder(str_trunc(feature,w=50), -R2_evo),
                   x = R2_evo, fill=categories)) +
      geom_bar(stat='identity',position='dodge')  + 
        #geom_text(aes(label=feat_n),hjust='inward',size=3,col='black') +
        facet_wrap(~source,scales = 'free_y',ncol=5) +
        theme( axis.text = element_text(size=12),legend.position = 'top') + scale_fill_tableau() +
      xlab('Correlation with Residual Evo.Rate') + ylab('Features') +
      guides(y = guide_axis(check.overlap=F, n.dodge=1.5))
plot(F4)
save_plot(plot = F4, 'output/F4.evorate-residuals-features.png', base_height=20, base_width=25)


categories=unique(test$categories)
for( icat in seq_along(categories) ){
  CAT = categories[icat]
  S4 = ggplot(test %>% dplyr::filter(categories == CAT & feat_n>100),# %>% slice_min_max( order_by=.res_evo.SR.avg, n=10 ), 
               aes(y = reorder(str_trunc(feature,w=50), -R2_evo),
                   x = R2_evo, fill=source)) +
      geom_bar(stat='identity',position='dodge')  + 
        geom_text(aes(label=feat_n),hjust='inward',size=3,col='black') + 
        facet_wrap(~categories,scales = 'free_y')
  plot(S4)
  save_plot(path="output/", filename = sprintf('S4_%s.features-residuals-%s.png',icat,CAT), plot = S4,device = 'png', scale = 2 )
}


feat.fit <-   feat.rk %>% 
             nest(data = -c(col_feat, feat_n)) %>% 
               mutate(
                 fit = map(data, ~ lm(rk.value  ~ rk.mpc,data=.x)),
                 augmented = map(fit, augment) ) %>%
               unnest(cols=augmented ) 
#fit.ER = inner_join(fit.ppm.ER,fit.mpc.ER, by=nonfit.col, suffix=c('.ppm','.mpc'))

```


```{r transform-}
propsub = propfit %>% dplyr::filter( property  %in% c('essential_core') )

p1= ggplot(evo_mpc.fit, aes(y = EVO.FULL, x = MPC)) +
          geom_ribbon(aes(ymin=.fitted.evo+0.01, ymax=Inf), fill='#BB0033',alpha=0.3)+
          geom_ribbon(aes(ymin=-Inf,ymax=.fitted.evo-0.01), fill='#00BB33',alpha=0.3)+
          geom_abline(slope=0,intercept = mean(evo_mpc.fit$EVO.FULL),size=1,col='gray',linetype=2)+
          geom_point(fill='lightgray',shape=19,alpha=0.3) +
          geom_segment(aes(xend = MPC, yend = .fitted.evo), col='gray',linetype='12',size=0.3,alpha=0.5) +
          # Overlay the protein sharing a particular properties
          geom_line(mapping=aes(y=.fitted.evo),size=2) +
          geom_segment(data=propsub, aes(xend = MPC, yend = .fitted.evo,color=property),linetype=2,size=0.5,alpha=0.8) +
          geom_point(data=propsub, aes(color=property), shape=19,size=1.4) + scale_color_simpsons() +
          annotate("text",y = 3, x=0.8, label='Over-estimated',col='#BB0033',vjust='inward',hjust='inward',size=8)+
          annotate("text",y = 0, x=0.8, label='Under-estimated',col='#00BB33',vjust='inward',hjust='inward',size=8)+
          xlab('Protein abundance (log10 mpc)') + ylab('Mean ER (full seq.)') + theme(legend.position = 'top',legend.direction = 'vertical')

plot(p1)

p2=     ggplot(data=evo_mpc.fit, aes(y=.resid.evo,x=MPC)) + 
        geom_ribbon(aes(ymin=0.01, ymax=Inf), fill='#BB0033',alpha=0.3)+
        geom_ribbon(aes(ymin=-Inf,ymax=-0.01), fill='#00BB33',alpha=0.3)+
        geom_abline(slope=0,intercept = 0,size=1)+
        geom_point(col='gray30',shape=19,alpha=0.3) +
        geom_segment(aes(xend = MPC, yend = 0),col='gray',linetype='12',size=0.3,alpha=0.5) +
        geom_segment(data=propsub,aes(xend = MPC, y=0, yend = .resid.evo,color=property),linetype=2,size=0.5,alpha=0.8) +
        geom_point(data=propsub, aes(color=property), shape=19,size=1.4) + scale_color_simpsons()+
        annotate("text",y = 2, x=0.8, label='Over-estimated',col='#BB0033',vjust='inward',hjust='inward',size=8)+
        annotate("text",y = -1.2, x=0.8, label='Under-estimated',col='#00BB33',vjust='inward',hjust='inward',size=8)+
        xlab('Protein abundance (log10 mpc)') + ylab('Residuals (ER)') + theme(legend.position="none")
plot(p2) 

propsubres = .res_prop %>% dplyr::filter( property  %in% c('essential_core') )
YMIN = min(-0.2,propsubres$.res_evo.SR.avg)%>% round(digits = 1)
YMAX = max(0.2,propsubres$.res_evo.SR.avg) %>% round(digits = 1)
p3 =  ggplot(propsubres,
             aes(x = reorder(str_trunc(property,w=50), -.res_evo.SR.avg),
                 y = .res_evo.SR.avg, fill=property), group=source) +
            geom_hline(yintercept = 0,size=1,col='gray',linetype=2) +
            geom_rect(aes(xmin=-Inf,xmax=Inf,ymin=0,ymax=Inf), fill='#BB0033',alpha=0.3)+
            geom_rect(aes(xmin=-Inf,xmax=Inf,ymin=-Inf,ymax=0), fill='#00BB33',alpha=0.3)+
            annotate("text",y = Inf, x=Inf, label='Over-estimated',col='#BB0033',vjust='inward',hjust='inward',size=5)+
            annotate("text",y = -Inf, x=Inf, label='Under-estimated',col='#00BB33',vjust='inward',hjust='inward',size=5)+
            geom_bar(stat='identity',position='dodge') +
            geom_text(aes(label=N),vjust='inward',size=8) + 
            theme( axis.text = element_text(size=14)) + 
            ylim(YMIN,YMAX) + scale_fill_simpsons() +  coord_flip()+  theme(legend.position="none",axis.text.y = element_text(angle=45))+
            ylab("Average residual evolutionary rate") + 
            xlab('Property')
plot(p3)
save_plot(plot =  plot_grid(p1,p2,p3,ncol = 3), filename = 'output/transformation-to-residuals.png',base_width=15,base_height = 5)
```

```{r F3-yeast-QR-code }
# n_distinct(prop$GNAME)/n_distinct(prop$property)
# prop %>% filter(has_ortho) %>% ungroup() %>% summarise(np=n_distinct(property), ng=n_distinct(GNAME), r=ng/np)
# prop %>% ungroup() %>% summarise(np=n_distinct(property), ng=n_distinct(GNAME), r=ng/np)
# 
# data.to.f1  = prop %>% mutate(propinfo = factor(has_prop+2*has_ortho, levels = c(0,1,2,3), labels('none','nocons_prop','cons_prot','cons_prop')))
# f1.1 = ggplot(data.to.f1) +
#   geom_raster(aes(x=GNAME,y=property,fill=propinfo),show.legend = F) +
#   scale_fill_manual(values=c('white','black')) +
#   scale_fill_manual(values=c('none'='white','nocons_prop'='red','cons_prot'='gray','cons_prop'='black'),na.value='white') +
#   theme(legend.title = element_blank(), axis.text = element_blank())
# f1.2 = ggplot(data.to.f1 %>% filter(has_ortho)) +
#   geom_raster(aes(x=GNAME,y=property,fill=has_prop),show.legend = F,interpolate=T) +
#   scale_fill_manual(values=c('white','black')) +
#   theme(legend.title = element_blank(), axis.text = element_blank())
# f1 = plot_grid(f1.1,f1.2)
# save_plot(path="output/", filename = 'F1.properties-proteome.png', plot = f1,device = 'png', base_height = 13,base_width=26)
# save_plot(path="output/", filename = 'F1.properties-proteome.pdf', plot = f1,device = 'pdf', base_height = 13,base_width=26)
# # pheatmap::pheatmap(prop[,-1]*1, 
#                    show_rownames = T,show_colnames = F,
#                    cluster_rows = T, cluster_cols = F, 
#                    cellwidth = 25, cellheight = 5,
#                    border_color = NA, color = c('black','white'), 
#                    filename='output/heatmap-protein-binary-properties.pdf'
#                    )

#library(proxy)
# Example of Jaccard distance on binary data
# x <- matrix(sample(c(FALSE, TRUE), 50, rep = TRUE), ncol = 10)
# dist(x, x, method = "Jaccard") # cross distances


# hc = hclust(d=stats::as.dist(prop_jacc),method = 'ward.D2')
# grp = cutree(hc,k=100) 

# pheatmap::pheatmap(prop_jacc,
#                    show_colnames = F, show_rownames = T,
#                    cluster_cols = hc, cluster_rows = hc,
#                    clustering_distance_rows = prop_jacc,
#                    clustering_distance_cols = prop_jacc,
#                    cutree_rows = 100,cutree_cols = 100,
#                    fontsize=3, border_color = NA, silent=T, 
#                    cellheight=3, cellwidth=3,
#                    filename="output/properties-jaccard.pdf", height=50, width=50)
# which(prop_jacc < 0.5,arr.ind=T)
# # Number of clusters
# library(factoextra)
# library(NbClust)
# m.prop_jacc=base::as.matrix(stats::as.dist(prop_jacc))
# 
# caret::findCorrelation
# 
# brk.minor=seq(5,45,by=5)
# brk.major=seq(50,50,by=25)
```



## Compute properties similarities

To compare properties, we will use the Jaccard distance on the binary matrix where rows are proteins and columns are the properties. The distance corresponds to the proportion of (TRUE, TRUE) pairs, but not considering (FALSE, FALSE) pairs.

So it compares the intersection with the union of protein sets between properties.


```{r properties-similarity}
# Pearson correlation between properties
PROP_cor = preload("output/properties-correlation.rds",
                   coop::pcor(PROP.m), 
                   doing="Computing Pearson correlation between properties...")
# Jaccard distance between properties
PROP_jacc = preload("output/properties-jaccard.rds", 
                   proxy::dist(PROP.m, PROP.m, by_rows=F, method="Jaccard"),
                   doing="Computing Jaccard distance between properties...") 

FEAT_scor = preload("output/features-spearman.rds",
                   cor(FEAT.m,use='pairwise.complete.obs',method = 'spearman'), 
                   doing="Computing Spearman correlation between features...")

```

