---
title: "Protein Categories"
author: "Benjamin Dubreuil"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("src/utils.r")
source("src/function_annotation.r")
source("src/function_sequence.r")
source("src/function_phylogenetic.r")
source("src/function_analysis.r")
source("src/function_datalocal.r")
source("src/function_datapub.r")
library(tictoc)
library(tidyverse)
```

To understand the organization of the yeast proteome, he factors that influence its evolution?
A concise answer designates protein expression as the driver for molecular evolution.
Indeed, highly expressed proteins are the most conserved.
the influence on its organization, we would like to 

- Protein categories

Each category corresponds to a set of protein identifiers which share a common non-quantitative trait (e.g. localization, pathway, function...)

- Protein features

A feature is a quantitative measure related to proteins parameter (e.g. disorder content, sequence length)

## Reference proteome sequence

```{r reference}
# REFERENCE PROTEIN DATABASE
# Saccharomyces genome database (SGD)
SGD  = load.sgd.features()
# Universal protein resource (UNIPROT)
file.uniprot.feature="data/uniprot-features.rds"
if( !file.exists(file.uniprot.feature) ){
  saveRDS(load.uniprot.features(),file.uniprot.feature) # Takes >250sec for ~6700 ids
}

# REFERENCE PROTEOME SEQUENCES
S288C  = load.sgd.proteome(withORF=T)
CDS    = load.sgd.CDS(withORF = T)
UNI.SC = load.uniprot.proteome('yeast')

UNIPROT = readRDS(file.uniprot.feature) %>%
          group_by(UNIPROTKB) %>%
          mutate( is_uniref = UNIPROTKB %in% names(UNI.SC), one2one = (n()==1) ) %>% 
          left_join(SGD %>% dplyr::select(sgdid,name,qual,gname,chr), by = c("SGD" = 'sgdid')) %>%
          dplyr::rename(ORF = name, GNAME=gname, CHR=chr)

# CROSS-REFERENCE PROTEOME
PROTEOME = UNIPROT %>% dplyr::select(UNIPROTKB, SGD, ORF, GNAME, PNAME)
CATEGORIES = PROTEOME
```

## Sequence features
```{r genetic}
has_intron = SGD %>% filter(type=='intron') %>% pull(parent) %>% intersect(y=names(S288C))
CATEGORIES$has_intron = CATEGORIES$ORF %in% has_intron

orf.len = tibble(orf=names(S288C),len=width(S288C))
Qlen=quantile(orf.len$len,seq(0,1,0.01))
orf.gc  = tibble(orf=names(CDS),pGC = rowSums(letterFrequency(CDS, letters="CG",as.prob = T)))
QGC = quantile(orf.gc$pGC,seq(0,1,0.01))

longest  = orf.len %>% filter(len > Qlen['75%']) %>% pull(orf)
shortest = orf.len %>% filter(len < Qlen['25%']) %>% pull(orf)
average  = orf.len %>% filter(between(len,Qlen['25%'],Qlen['75%'])) %>% pull(orf)
midGC  = orf.gc %>% filter(between(pGC,QGC['25%'],QGC['75%'])) %>% pull(orf)
loGC   = orf.gc %>% filter(pGC<QGC['25%']) %>% pull(orf)
hiGC   = orf.gc %>% filter(pGC>QGC['75%']) %>% pull(orf)


CAT$longest = CAT$ORF %in% longest
CAT$average = CAT$ORF %in% average
CAT$shortest = CAT$ORF %in% shortest
CAT$midGC = CAT$ORF %in% midGC
CAT$lowGC = CAT$ORF %in% loGC
CAT$highGC = CAT$ORF %in% hiGC

```

## Biophysical features
```{r biophysical}
AAFREQ = letterFrequency(UNI.SC,as.prob = T,letters = get.AA1())
AABIN = apply(AAFREQ,2,quantile,probs=seq(0,1,len=11))
pheatmap::pheatmap(AABIN,cluster_rows = F)

DUB = load.dubreuil2019.data(3)
NDOM = DUB %>%
        group_by(id) %>%
        count(pfam_ids, name='n.pfam')


```

## Interactions features
```{r interactions}
# PROTEIN-PROTEIN INTERACTIONS
#/media/elusers/users/hugo/07_3DComplex_scripts/scripts_PPIs_networks/stack_studies_PPIs_nored_PMID_28122020_INTACT.R
PPI = read.csv("/media/elusers/users/hugo/07_3DComplex_scripts/PPIs_analysis/INTACT/PPIs_cerevisiae.txt",sep = "\t", quote = "", stringsAsFactors = F)
colnames(PPI) = c("protA","protB","altA","altB","aliasA","aliasB","method","pub.author1","pub.id","taxA","taxB","typeAB","src.db","int.id","conf","expansion.method","bio.roleA","bio.roleB","exp.roleA","exp.roleB","typeA","typeB","xrefA","xrefB","xrefAB","annotA","annotB","annotAB","host","parAB","created","updated","checksumA","checksumB","checksumAB","negative","featA","featB","stoichioA","stoichioB","id.met.protA","id.met.protB")
# remove duplicated interactions (with different Pubmed ID)
# 
# keep only protein interactions (uniprot/orf)
# filter for direct method
# return the matrix of interactions

dplyr::select(-c(created,updated,checksumA,checksumB,checksumAB,taxA,taxB))

# COMPLEXES
CYC = load.pu.2008.data() # BASED ON ORF
CPX = load.meldal.2019.data() # BASED ON UNIPROT
test = CYC %>% filter(n_members > 5) %>% pivot_wider(names_from = Complex,values_from = c(ORF))
test[11:50,]
```

## Phenotypes features
```{r phenotype}
# Essentiality
ESS = load.vanleeuwen2020.data()
essential = ESS %>% 
                filter(disp !='Dispensable') %>% 
                filter(disp == 'Core' | (KO_exp =='Indispensable' & disp_score < 0.4 )) %>%
                pull(ORF)

# Dispensability
dispensable = ESS %>% filter(disp =='Dispensable' ) %>% pull(ORF)

CAT$phenotype.essential =  CAT$ORF %in% essential
CAT$phenotype.dispensable =  CAT$ORF %in% dispensable

table(ESS$disp[ESS$ORF %in% CAT$dispensable])
table(ESS$disp[ESS$ORF %in% CAT$essential])

CHM = load.lee2014.data() # Chemogenomic fitness signatures
# MDR -> multi-drug resistance (>20% fitness)

```

## Subcellular localization
```{r localization}
UNIGO = get.uniprot.go(UNIPROT$UNIPROTKB)
UNILOC = get.uniprot.localization(annot=UNIPROT,loc_to_columns = T)
uniloc = get.uniprot.localization(annot=UNIPROT,loc_to_columns = F)

CAT$has_foci = UNILOC %>% filter(HAS_FOCI) %>% pull(id)
CAT$has_iso = UNILOC %>% filter(HAS_ISOFORM) %>% pull(id)

THR.LOC.SIZE = 50 # MINIMUM NUMBER OF PROTEINS PER COMPARTMENT 
LOC.size = sort(colSums(UNILOC[,-c(1:3)]))
LOC50 = sort( names( LOC.size[LOC.size>THR.LOC.SIZE] ) )

#writeLines(sprintf("CAT$loc.%s = UNILOC %%>%% filter(%s == 1) %%>%% pull(id)",LOC50,LOC50))
CAT$loc.bud_neck = UNILOC %>% filter(bud_neck == 1) %>% pull(id)
CAT$loc.cell_membrane = UNILOC %>% filter(cell_membrane == 1) %>% pull(id)
CAT$loc.cell_wall = UNILOC %>% filter(cell_wall == 1) %>% pull(id)
CAT$loc.centromere = UNILOC %>% filter(centromere == 1) %>% pull(id)
CAT$loc.chromosome = UNILOC %>% filter(chromosome == 1) %>% pull(id)
CAT$loc.cytoplasm = UNILOC %>% filter(cytoplasm == 1) %>% pull(id)
CAT$loc.cytoplasmic_side = UNILOC %>% filter(cytoplasmic_side == 1) %>% pull(id)
CAT$loc.cytoskeleton = UNILOC %>% filter(cytoskeleton == 1) %>% pull(id)
CAT$loc.endoplasmic_reticulum_membrane = UNILOC %>% filter(endoplasmic_reticulum_membrane == 1) %>% pull(id)
CAT$loc.endosome_membrane = UNILOC %>% filter(endosome_membrane == 1) %>% pull(id)
CAT$loc.golgi_apparatus = UNILOC %>% filter(golgi_apparatus == 1) %>% pull(id)
CAT$loc.golgi_apparatus_membrane = UNILOC %>% filter(golgi_apparatus_membrane == 1) %>% pull(id)
CAT$loc.gpi.anchor = UNILOC %>% filter(gpi.anchor == 1) %>% pull(id)
CAT$loc.lipid.anchor = UNILOC %>% filter(lipid.anchor == 1) %>% pull(id)
CAT$loc.membrane = UNILOC %>% filter(membrane == 1) %>% pull(id)
CAT$loc.mitochondrion = UNILOC %>% filter(mitochondrion == 1) %>% pull(id)
CAT$loc.mitochondrion_inner_membrane = UNILOC %>% filter(mitochondrion_inner_membrane == 1) %>% pull(id)
CAT$loc.mitochondrion_intermembrane_space = UNILOC %>% filter(mitochondrion_intermembrane_space == 1) %>% pull(id)
CAT$loc.mitochondrion_matrix = UNILOC %>% filter(mitochondrion_matrix == 1) %>% pull(id)
CAT$loc.mitochondrion_membrane = UNILOC %>% filter(mitochondrion_membrane == 1) %>% pull(id)
CAT$loc.mitochondrion_outer_membrane = UNILOC %>% filter(mitochondrion_outer_membrane == 1) %>% pull(id)
CAT$loc.multi.pass_membrane_protein = UNILOC %>% filter(multi.pass_membrane_protein == 1) %>% pull(id)
CAT$loc.nucleolus = UNILOC %>% filter(nucleolus == 1) %>% pull(id)
CAT$loc.nucleus = UNILOC %>% filter(nucleus == 1) %>% pull(id)
CAT$loc.nucleus_membrane = UNILOC %>% filter(nucleus_membrane == 1) %>% pull(id)
CAT$loc.peripheral_membrane_protein = UNILOC %>% filter(peripheral_membrane_protein == 1) %>% pull(id)
CAT$loc.secreted = UNILOC %>% filter(secreted == 1) %>% pull(id)
CAT$loc.single.pass_membrane_protein = UNILOC %>% filter(single.pass_membrane_protein == 1) %>% pull(id)
CAT$loc.single.pass_type_ii_membrane_protein = UNILOC %>% filter(single.pass_type_ii_membrane_protein == 1) %>% pull(id)
CAT$loc.vacuole_membrane = UNILOC %>% filter(vacuole_membrane == 1) %>% pull(id)

multiloc = UNILOC %>% 
            group_by(id) %>% 
            mutate( nloc = sum(c_across(!starts_with("HAS_TO"))),
                    nloc50 = sum(c_across(LOC50)),
                    localized = paste(sort(uniloc$loc[ uniloc$id == id ]), collapse = "#"),
                    localized50 = paste(intersect(LOC50,sort(uniloc$loc[ uniloc$id == id ])), collapse = "#")) %>%
            filter(nloc50>3) %>% arrange(nloc50,nloc) %>%
            dplyr::select('id','HAS_FOCI','HAS_ISOFORM','nloc','nloc50','localized50')
  
head(multiloc)

```

```{python}


```

```{r categories-table}
# Make categories table with:

```