---
title: "Protein Categories"
author: "Benjamin Dubreuil"
date: "January 6, 2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
source("src/utils.r")
source("src/function_annotation.r")
source("src/function_sequence.r")
source("src/function_phylogenetic.r")
source("src/function_analysis.r")
source("src/function_datalocal.r")
source("src/function_datapub.r")
library(tictoc)
library(tidyverse)

CAT=list()
```

## Reference proteome sequence
```{r reference}
# MAIN PROTEIN FEATURES
SGD  = load.sgd.features()

if( !file.exists("data/uniprot-features.rds") ){
  saveRDS(load.uniprot.features(),"data/uniprot-features.rds") # Takes >250sec for ~6700 ids
}
UNIPROT = readRDS("data/uniprot-features.rds")
#UNIPROT =  
# PRIMARY REFERENCE SEQUENCES
S288C  = load.sgd.proteome(withORF=T)
CDS    = load.sgd.CDS(withORF = T)
UNI.SC = load.uniprot.proteome('yeast')

UNIPROT = UNIPROT %>%
          group_by(UNIPROTKB) %>%
          mutate( is_uniref = UNIPROTKB %in% names(UNI.SC), one2one = (n()==1) ) 

orf.len = tibble(orf=names(S288C),len=width(S288C))
Qlen=quantile(orf.len$len,seq(0,1,0.01))
orf.gc  = tibble(orf=names(CDS),pGC = rowSums(letterFrequency(CDS, letters="CG",as.prob = T)))
QGC = quantile(orf.gc$pGC,seq(0,1,0.01))
```



## Genetic features
```{r genetic}
CAT$has_intron = SGD   %>% filter(type=='intron')    %>% pull(parent)

CAT$longest  = orf.len %>% filter(len > Qlen['75%']) %>% pull(orf)
CAT$shortest = orf.len %>% filter(len < Qlen['25%']) %>% pull(orf)
CAT$average  = orf.len %>% filter(between(len,Qlen['25%'],Qlen['75%'])) %>% pull(orf)

CAT$midGC  = orf.gc %>% filter(between(pGC,QGC['25%'],QGC['75%'])) %>% pull(orf)
CAT$loGC   = orf.gc %>% filter(pGC<QGC['25%']) %>% pull(orf)
CAT$hiGC   = orf.gc %>% filter(pGC>QGC['75%']) %>% pull(orf)
```

## Biophysical features
```{r biophysical}
letterFrequency(UNI.SC,as.prob = T,letters = get.AA1())
DUB = load.dubreuil2019.data(3)
head(DUB)


```

## Interactions features
```{r interactions}
#/media/elusers/users/hugo/07_3DComplex_scripts/scripts_PPIs_networks/stack_studies_PPIs_nored_PMID_28122020_INTACT.R

#PPI = read.csv("/media/elusers/users/hugo/07_3DComplex_scripts/PPIs_analysis/INTACT/PPIs_cerevisiae.txt",sep = "\t", quote = "", stringsAsFactors = F)

head(PPI)

# COMPLEXES
EBI.complextab = read.delim("ftp://ftp.ebi.ac.uk/pub/databases/intact/complex/current/complextab/559292.tsv",
                            stringsAsFactors = F,header = T, sep = '\t', 
                            col.names = c('CPLX_ID','CPLX_NAME','CPLX_ALIAS','TAX_ID',
                                          'MEMBERS_STOCHIO','CONF','EXP_EVID','GO_ANNOT',
                                          'CROSSREF','DESC',
                                          'CPLX_PROP','CPLX_ASSEMBLY',
                                          'LIG','DISEASE','AGONIST','ANTAGONIST',
                                          'COMMENT','SOURCE','EXPANDED_MEMBERS'))

CX=EBI.complextab %>% dplyr::select("CPLX_NAME", 
                                    "MEMBERS_STOCHIO", "EXPANDED_MEMBERS",
                                    "CONF","EXP_EVID","CPLX_PROP","CPLX_ASSEMBLY",
                                    "LIG","DISEASE","AGONIST","ANTAGONIST") %>%
                   mutate( members = strsplit(EXPANDED_MEMBERS,split="\\|") )

```

## Phenotypes features
```{r phenotype}
# Essentiality
# Dispensability
```

## Subcellular localization
```{r localization}
UNIGO = get.uniprot.go(UNIPROT$UNIPROTKB)
UNILOC = get.uniprot.localization(annot=UNIPROT,loc_to_columns = T)
uniloc = get.uniprot.localization(annot=UNIPROT,loc_to_columns = F)

CAT$has_foci = UNILOC %>% filter(HAS_FOCI) %>% pull(id)
CAT$has_iso = UNILOC %>% filter(HAS_ISOFORM) %>% pull(id)

THR.LOC.SIZE = 50 # MINIMUM NUMBER OF PROTEINS PER COMPARTMENT 
LOC.size = sort(colSums(UNILOC[,-c(1:3)]))
LOC50 = sort( names( LOC.size[LOC.size>THR.LOC.SIZE] ) )

#writeLines(sprintf("CAT$loc.%s = UNILOC %%>%% filter(%s == 1) %%>%% pull(id)",LOC50,LOC50))
CAT$loc.bud_neck = UNILOC %>% filter(bud_neck == 1) %>% pull(id)
CAT$loc.cell_membrane = UNILOC %>% filter(cell_membrane == 1) %>% pull(id)
CAT$loc.cell_wall = UNILOC %>% filter(cell_wall == 1) %>% pull(id)
CAT$loc.centromere = UNILOC %>% filter(centromere == 1) %>% pull(id)
CAT$loc.chromosome = UNILOC %>% filter(chromosome == 1) %>% pull(id)
CAT$loc.cytoplasm = UNILOC %>% filter(cytoplasm == 1) %>% pull(id)
CAT$loc.cytoplasmic_side = UNILOC %>% filter(cytoplasmic_side == 1) %>% pull(id)
CAT$loc.cytoskeleton = UNILOC %>% filter(cytoskeleton == 1) %>% pull(id)
CAT$loc.endoplasmic_reticulum_membrane = UNILOC %>% filter(endoplasmic_reticulum_membrane == 1) %>% pull(id)
CAT$loc.endosome_membrane = UNILOC %>% filter(endosome_membrane == 1) %>% pull(id)
CAT$loc.golgi_apparatus = UNILOC %>% filter(golgi_apparatus == 1) %>% pull(id)
CAT$loc.golgi_apparatus_membrane = UNILOC %>% filter(golgi_apparatus_membrane == 1) %>% pull(id)
CAT$loc.gpi.anchor = UNILOC %>% filter(gpi.anchor == 1) %>% pull(id)
CAT$loc.lipid.anchor = UNILOC %>% filter(lipid.anchor == 1) %>% pull(id)
CAT$loc.membrane = UNILOC %>% filter(membrane == 1) %>% pull(id)
CAT$loc.mitochondrion = UNILOC %>% filter(mitochondrion == 1) %>% pull(id)
CAT$loc.mitochondrion_inner_membrane = UNILOC %>% filter(mitochondrion_inner_membrane == 1) %>% pull(id)
CAT$loc.mitochondrion_intermembrane_space = UNILOC %>% filter(mitochondrion_intermembrane_space == 1) %>% pull(id)
CAT$loc.mitochondrion_matrix = UNILOC %>% filter(mitochondrion_matrix == 1) %>% pull(id)
CAT$loc.mitochondrion_membrane = UNILOC %>% filter(mitochondrion_membrane == 1) %>% pull(id)
CAT$loc.mitochondrion_outer_membrane = UNILOC %>% filter(mitochondrion_outer_membrane == 1) %>% pull(id)
CAT$loc.multi.pass_membrane_protein = UNILOC %>% filter(multi.pass_membrane_protein == 1) %>% pull(id)
CAT$loc.nucleolus = UNILOC %>% filter(nucleolus == 1) %>% pull(id)
CAT$loc.nucleus = UNILOC %>% filter(nucleus == 1) %>% pull(id)
CAT$loc.nucleus_membrane = UNILOC %>% filter(nucleus_membrane == 1) %>% pull(id)
CAT$loc.peripheral_membrane_protein = UNILOC %>% filter(peripheral_membrane_protein == 1) %>% pull(id)
CAT$loc.secreted = UNILOC %>% filter(secreted == 1) %>% pull(id)
CAT$loc.single.pass_membrane_protein = UNILOC %>% filter(single.pass_membrane_protein == 1) %>% pull(id)
CAT$loc.single.pass_type_ii_membrane_protein = UNILOC %>% filter(single.pass_type_ii_membrane_protein == 1) %>% pull(id)
CAT$loc.vacuole_membrane = UNILOC %>% filter(vacuole_membrane == 1) %>% pull(id)

multiloc = UNILOC %>% 
            group_by(id) %>% 
            mutate( nloc = sum(c_across(!starts_with("HAS_TO"))),
                    nloc50 = sum(c_across(LOC50)),
                    localized = paste(sort(uniloc$loc[ uniloc$id == id ]), collapse = "#"),
                    localized50 = paste(intersect(LOC50,sort(uniloc$loc[ uniloc$id == id ])), collapse = "#")) %>%
            filter(nloc50>3) %>% arrange(nloc50,nloc) %>%
            dplyr::select('id','HAS_FOCI','HAS_ISOFORM','nloc','nloc50','localized50')
  
head(multiloc)

```
