---
title: "Yeast Proteome Evolution"
date: "`r format(Sys.time(), 'Last Modified: %d-%b-%Y %R %p')`"
subtitle: "Integrating timescales in protein evolution from distant species to populations"
author: "[Benjamin Dubreuil](https://github.com/benjamin-elusers/)"
mail: "benjamin.dubreuil@weizmann.ac.il"
github: "benjamin-elusers"
home: "https://github.com/benjamin-elusers/yeastomics/"
output:
  rmdformats::downcute:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: true
    highlight: tango
#  html_document: 
    code_download: true
    code_folding: hide
    fig_caption: yes
    keep_md: yes
#    toc: yes
    toc_depth: 3
    toc_float:
      toc_collapsed: true
    number_sections: true
    theme: lumen
  editor_options: 
    chunk_output_type: console
  bibliography: bibliography.bib
  csl: biomed-central.csl
---

<!-- Github Logo -->
<a href="https://github.com/benjamin-elusers/yeastomics/" class="github-corner" aria-label="View source on Github">
  <svg width="80" height="80" viewBox="0 0 250 250" style="fill:#69b3a2; color:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true">
    <path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path>
    <path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path>
    <path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path>
  </svg>
</a>

<style> 
.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}
@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}
@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}
</style>

```{r setup, include=F, echo=F, warnings=F, message=F}
knitr::opts_chunk$set(echo = T, collapse = T, cache.lazy = F,autodep = T)
suppressPackageStartupMessages({
  library(xfun)
  # Load dependencies package (optional) ---------------------------------------
  dep.pkg = c("BiocGenerics","Biobase","S4Vectors","parallel","stats4","IRanges",
              "XVector","hutils","AnnotationDbi")
  xfun::pkg_attach2(dep.pkg)
  # Load graphical packages ----------------------------------------------------
  gr.pkg = c("ggplot2","gridExtra","ggthemes","ggsci","ggpubr","ggrepel","plotly")
  xfun::pkg_attach2(gr.pkg)
  # Load required packages -----------------------------------------------------
  main.pkg = c("tictoc","tidyverse","Biostrings","hablar")
  xfun::pkg_attach2(main.pkg)
  # Load custom local functions (available via Github repo) --------------------
  source("src/utils.r")
  source("src/function_annotation.r")
  source("src/function_sequence.r")
  source("src/function_phylogenetic.r")
  source("src/function_analysis.r")
  source("src/function_datalocal.r")
  source("src/function_datapub.r")
  source("src/function_YK11.r")
  source("src/theme_black.r")

})
# turn off annoying messages from dplyr::summarise
options(dplyr.summarise.inform = FALSE)

preload = function(saved.file,loading.call,doing='create data...'){
  library(tictoc)
  if( !file.exists(saved.file) ){
    cat(doing)
    tic(doing)
    res = eval(substitute(loading.call))
    saveRDS(res,saved.file)
    toc()
  }else{
    res = readRDS(saved.file)
  }
  return(res)
}

#theme_set(theme_black())
theme_set(theme_bw(base_size=16,base_line_size = 0.5) +theme(plot.background = element_blank()))

MAIN.COLOR = 'black'
```

# Reference proteomes {.tabset .tabset-pills}

## SGD

_Saccharomyces Genome Database_

The reference genome sequences of the widely used Baker's yeast are maintained by
tbe [SGD](www.yeastgenome.org), corresponding to the strain **S. cerevisiae S288C**.

The SGD also gather many key information such as:

- genome assembly
- Protein-coding sequence
- gene/protein nomenclature
- biological activity (interaction/regulation/complexes)
- curated annotations and descriptins (gene ontology)
- gene and protein expression
- accurate scientific literature,
- genotype/phenotype experiments
...

```{r sgd}
# YEAST REFERENCE SEQUENCES
## SGD
S288C = load.sgd.proteome(withORF=T,rm.stop=F) # Reference SGD protein sequences
S288C.cds = load.sgd.CDS(withORF=T) # Reference coding sequences (CDS)
SGD = load.sgd.features() # main SGD gene/protein features
```

## UniProt

_Universal Protein resource_

[UniProt](www.uniprot.org) provides the most comprehensive resource of protein 
sequence and functional information.
<br>
Most protein features are typically stored on the database called UniProt 
KnowledgeBase (UniProtKB).
<br>
This database is a major hub for many specialized databases and webservers that
helps understand proteins role.
<br>
Their [reference proteome](https://www.uniprot.org/uniprot/?query=proteome:UP000002311)
gather proteins for which expression or proof of physical existence has been shown.

```{r unip}
## UNIPROT
# saveRDS(load.uniprot.features(),"data/uniprot-features.rds") # Takes >250sec for ~6700 ids
UNI.SC = load.uniprot.proteome('yeast') # Reference UniprotKB sequences (Uniref)
UNI = readRDS("data/uniprot-features.rds") %>% # main UNIPROT gene/protein features
        group_by(UNIPROTKB) %>%
        mutate( is_uniref = UNIPROTKB %in% names(UNI.SC), one2one = (n()==1) ) 
```

## YK11

_1,011 Saccharomyces cerevisiae isolates_

The [1002 Yeast Genomes project](http://1002genomes.u-strasbg.fr/) constitutes 
the most comprehensive genomic dataset on a single species of yeast. Ultimately 
provide the most extensive view of the genetic and phenotypic diversity within
this model species to date.
<br>
The project was originally slated to include 1002 strains from diverse global locations,
as well as a variety of ecological sources. A total of 1011 whole genome sequences
have been produced and published in [*J. Peter et al.*, Nature 2018](https://doi.org/10.1038/s41586-018-0030-5).

```{r y1k11}
## 1011 strains
riboseq_strains = c('AMH','BAN','BED','BPL','BTI','CMP','CPI','CQC') # Strains with riboseq data (on 14/01/21)
strains.info = load.peter2018.data() %>%  # strains info from supp mat of Science paper
                 mutate( has_riboseq = standardized.name %in% riboseq_strains)

# saveRDS(load.1011.strains(),"data/proteome-1011-strains.rds") # Takes >5mn for 6575 ids
YK11 = readRDS("data/proteome-1011-strains.rds")  # 1011 strains proteomes sequences
```

All strains read were aligned on the reference yeast genome (ATCC 204508 / S288c). 
Therefore, all sequenced ORF should have exactly the same length as the references.

## Statistics

Below we calculate some general statistics on the reference yeast proteome.

```{r ref-id-map}
id_orf=names(S288C)
map_sgd2orf = SGD %>% dplyr::filter(name %in% id_orf) %>% dplyr::select(sgdid,name) 
id_sgd = map_sgd2orf$sgdid

UNI.1 = UNI # Backup of Uniprot features table
UNI   = left_join(UNI,map_sgd2orf, by=c('SGD'='sgdid')) %>% dplyr::rename(ORF=name)

id_uni = names(UNI.SC)
map_uni2sgd = UNI %>% dplyr::filter(UNIPROTKB %in% id_uni & one2one & !is.na(SGD)) %>% dplyr::select(UNIPROTKB,SGD,ORF) 
id_y1k = names(YK11)

n_sgd = length(S288C) %>% as.character()
n_uni = length(UNI.SC) %>% as.character()
n_y1k = length(YK11) %>% as.character()
```

Sequences of both SGD and UNIPROTKB proteomes are quite similar, yet, they don't
share the same number of proteins.
<br>
SGD contains **`r n_sgd`** sequences while UNIPROTKB has **`r n_uni`** sequences.

```{r ref-ali}
uniseq = UNI.SC[map_uni2sgd$UNIPROTKB]
sgdseq = rm.stop(S288C[map_uni2sgd$ORF])
aliref = align.pair.prot(uniseq, sgdseq) # Takes ~20s for 6042 ids
#saveRDS(pairwise.alignment.to.df(aliref),"data/sgduni-aligned-proteome.rds") 
# Takes ~8mn for 6042 ids
ALI_REF = readRDS("data/sgduni-aligned-proteome.rds")

ALI_STAT = group_by(ALI_REF,s1) %>%
           summarise( not_aligned=sum(!aligned), OL = mean(ol.12), PID = mean(pid.long) )
## OVERALL PROTEOME ALIGNMENT
# 99.93% (+/- 1.44%) identity on average for 6014 proteins
n_aligned = sum(ALI_STAT$PID==100) %>% as.character()
pid_avg = mean(ALI_STAT$PID) %>% round(2) %>% as.character()
pid_sd  = sd(ALI_STAT$PID) %>% round(2) %>% as.character()

## FOR MISALIGNED PROTEINS
n_misaligned = sum(ALI_STAT$PID<100) %>% as.character()
# 85.97% (+/- 16.10%) identity on average for 28 proteins
pid_avg.2 = mean(ALI_STAT$PID[ALI_STAT$not_aligned!=0]) %>% round(2) %>% as.character()
pid_sd.2 = sd(ALI_STAT$PID[ALI_STAT$not_aligned!=0]) %>%  round(2) %>% as.character()
n_unaligned = sum(ALI_STAT$not_aligned) %>% as.character()
```

On average, the mapped SGD/UNIPROT sequences share **`r pid_avg`%** identity **(+/- `r pid_sd`%)**
<br>
**`r n_aligned`** proteins are exactly identical (i.e. 100% identity) between the two proteomes.
<br>

**`r n_misaligned`** proteins contains mismatches/indels for at least **`r n_unaligned`** residues.
<br>
The misaligned proteins share on average **`r pid_avg.2`%** identity  **(+/- `r pid_sd.2`%)**

# Proteome datasets {.tabset .tabset-pills}

## Annnotations

_Gene ontology, SGD description, Uniprot commments_

The protein localizations can be obtained from curated annotations from 
Gene Ontology (GO) and the subcellular locations from Uniprot.

```{r load-annotations}

GO   = get.uniprot.go(id_uni)
CC   = GO %>% dplyr::filter(ONTOLOGY == CC) 
LOC.m  = get.uniprot.localization(annot = UNI,loc_to_columns = T)
LOC.df  = get.uniprot.localization(annot = UNI,loc_to_columns = F)
DESC = get.uniprot.sgd(id_uni) 
DESC.UNI = DESC %>% 
           left_join(UNI, by=c('UNIPROT'='UNIPROTKB','ORF'='ORF','SGD'='SGD'))

CAT.ORF=list()
CAT.ORF$has_intron = SGD %>% filter(type=='intron') %>% pull(parent) %>% intersect(id_orf)

ESS = load.vanleeuwen2020.data()
CAT.ORF$essential = ESS %>% 
                filter(disp !='Dispensable') %>% 
                filter(disp == 'Core' | (KO_exp =='Indispensable' & disp_score < 0.4 )) %>%
                pull(ORF)
CAT.ORF$dispensable = ESS %>% filter(disp =='Dispensable' ) %>% pull(ORF)

```

## Conservation

_Site-specific evolutionary rates_

[Rate4Site](https://www.tau.ac.il/~itaymay/cp/rate4site.html) is a program for 
detecting conserved amino-acid sites by computing the relative evolutionary rate
for each site in the multiple sequence alignment (MSA).
<br>
For the fungi lineage, we calculated the evolutionary rates of the orthologs 
defined in [*Wapinsky et al.*, Nature 2007](https://www.nature.com/articles/nature06107). 
<br>
For the S. cerevisiae population, we calculated the rate of evolution at 
every site of the 1011 isolates proteomes.This rate is expected to be better than
a non-synonymous SNP frequency. 
<br>
Indeed, the phylogenetic tree of the 1011 strains is taken into account to balance
for SNPs that occur in closely related strains relative to SNPs coming from the 
most recent common ancestor (MRCA).

```{r load-evorate}
WAP.r4s = load.wapinsky2007.data()   # Evolutionary rate for fungi lineage (~20s to load)
#saveRDS(load.rate4site_1011.data(), "data/rate4site-1011-strains.rds") (~3mn to load)
K11.r4s = readRDS("data/rate4site-1011-strains.rds") # Evolutionary rate for 1011 strains 
EVO = load.aligned.data(data.path="data/" ) # Protein dataset with aligned evolutionary rate

FUNGI = WAP.r4s %>%
        group_by(r4s_orf,r4s_size) %>%
        mutate(R4S_norm = R4S_res + abs(min_(R4S_res))) %>%
        summarise(r4s.fungi = mean_(R4S_norm),
                  max_r4s.fungi = max_(R4S_norm)
        )
POP  = K11.r4s %>% dplyr::filter( !is.na(ID) ) %>% 
       group_by(orf,len.s288c) %>%
       summarise(r4s.yk11 = mean_(SCORE),
                 max_r4s.yk11 = max_(SCORE)
       )
R4S = left_join(POP,FUNGI, by=c('orf'='r4s_orf', 'len.s288c'='r4s_size')) %>%
      mutate( log10.r4s.fungi = log10(r4s.fungi),
              log10.r4s.yk11 =log10(r4s.yk11))

CAT.ORF$ohnologs = load.byrne2005.data()$orf

# FIND OUTLIERS
# fit = lm(log10.r4s.fungi ~ log10.r4s.yk11, data = R4S) %>% broom::augment_columns(data=R4S,interval='confidence')
# cutoff.outlier= 20*mean(fit$.cooksd)
# R4S = fit %>% mutate(outlier = .cooksd > cutoff.outlier)

```

## Expression

Protein expression is taken from 
[*Ho et al.*, Cell Systems, 2018](https://doi.org/10.1016/j.cels.2017.12.004) 
which attempts to unify protein abundance datasets from the 
[Protein Abundance database](www.paxdb.org).
The unified protein expression is given in molecules per cell (MPC).
The coefficient of variation is calculated as:
<br>
$$\displaystyle CV = \frac{\sigma}{\mu}$$ 
<br> where: 
<br> sigma ($\sigma$) corresponds to the MPC standard deviation across datasets
<br> mu ($\mu$) corresponds is the MPC average across datasets

from all datasets is estimated corrected to account for bias 
due to each method (MS,GFP,TAP)


<br>
tRNA adapation index (tAI) is calculated from the coding sequence (CDS) from SGD as in 
[*Dos Reis et al.*, NAR, 2004](https://doi.org/10.1093/nar/gkh834).

Below is the crossed-correlations between all measurements of gene/protein expression:
```{r load-expression}
TAI = load.codon.usage(S288C.cds)
PAB = load.ho2018.data()

library(GGally)
gene_exp = c('tAI','mean.mpc','median.mpc','GFP.avg','MS.avg')
log10_exp = paste0("log10_",gene_exp[-1])
EXP = left_join(TAI,PAB, by=c('prot'='orf')) %>% 
        mutate( across(all_of(gene_exp[-1]), .fns = log10, .names = "log10_{col}" ) )

gg_point <- wrap(ggally_points, size = 0.5, color = MAIN.COLOR, alpha=0.3)
gg_cor <-  wrap(ggally_cor, title='r', method = "spearman", 
              color=MAIN.COLOR, bg='black', display_grid = FALSE)
F2 = ggpairs(data = EXP, 
             columns=c(gene_exp[1],log10_exp),
             upper = list(continuous = gg_cor),
             lower = list(continuous = gg_point),
             diag  = list(continuous = wrap("densityDiag", col=MAIN.COLOR, lwd=0.5))
            )
F2
      
```


# Yeast SNP {.tabset .tabset-pills}

Using the translated protein sequences from the 1011 isolates proteomes, I calculate
the frequency of amino acids at every site from each ORF across strains.
<br>
Some sites are ambiguous between strains, meaning that some strains may have an
alternative amino-acid at that position.
<br>


## SNP from 1011 strains

Since, I am looking at amino-acid frequencies we can only consider non-synonymous mutations.
<br>
Indels are not incorporated in the strains sequences.

```{r snp-proc}
SNP_FREQ=round(1/1011,4)
wt.file = 'data/YK11_WT.rds'
var.file = 'data/YK11_VAR.rds'
snppos.file = 'data/YK11_SNP_pos.rds'

P = tibble( orf=names(YK11), 
            n_strains=lengths(YK11),
            len = sapply(widths(YK11),unique)
    ) %>% left_join(get.width(S288C), by=c('orf'='orf'),suffix=c('','.s288c')) %>%
    mutate( match_wt = len == len.s288c )
VAR = preload(var.file, loading.call = map_df(YK11, get.variants,verbose=F),doing='Find all variants...')
SNP = get.variants.to.SNP(VAR) %>% 
        add_count(id,ref_pos,name='nvar') %>%
        group_by(id,ref_pos) %>% 
        mutate( alt_cumfr=sum(alt_fr)) %>% 
        left_join(P,by=c('id'='orf'))
WT = BS.to.df(S288C) %>% 
     group_by(orf) %>% 
     mutate(bin.pos=dplyr::ntile(wt_pos,100))

PROT_SNP = left_join(SNP,WT, by=c('id'='orf','ref_pos'='wt_pos')) %>%
      mutate(wt_low = alt_aa == wt_aa, 
             wt_missing=is.na(wt_aa),
             early_stop = (alt_aa == "*" & ref_pos != len), 
             dSTI.ref=get.score.mutation(ref_aa,alt_aa),
             dSTI.wt=get.score.mutation(wt_aa,alt_aa))
```

**SNP** is defined as a position containing at least **`r as.character(100*SNP_FREQ)`%**
alternative amino acids (*i.e variants*) across strains. 
<br>
The "wildtype" residue is the amino acid with the highest frequency across all strains.

A total of **`r TOTAL_SNP_VAR`** variants have been identified spread over
**`r TOTAL_SNP_POS`** positions of the proteome.

## SNP count

```{r snp-plot}

# COUNT ALL
singletons = PROT_SNP$nvar==1
earlystop = PROT_SNP$early_stop
TOTAL_SNP_VAR = nrow(PROT_SNP) %>% as.character()
TOTAL_SNP_POS = n_distinct(PROT_SNP$id,PROT_SNP$ref_pos) %>% as.character()
TOTAL_SNP_SINGLE = sum(singletons) %>% as.character()
TOTAL_SNP_MULTI = sum(!singletons) %>% as.character()
TOTAL_SNP_STOP = sum_(earlystop) %>% as.character()

BRK.FREQ = c(0,0.0001,1e-3,1e-2,5e-2,1e-1,5e-1)
LAB.FREQ = paste.pair(trimws(sprintf("%.4f",BRK.FREQ), whitespace = '0', which = 'right'),s='-')

PROT_SNP$fr_bin = cut(PROT_SNP$alt_fr,breaks = BRK.FREQ,labels = LAB.FREQ,ordered_result = T,include.lowest = T)
snp_count = PROT_SNP %>% 
            group_by(nvar) %>%
            summarize(var_count=sum(nvar)) %>% 
            ungroup() %>% mutate(var_mean = var_count / sum(var_count))

snp_byfreq =  PROT_SNP %>% 
              group_by(nvar,fr_bin) %>%
              summarize( var_count_byfreq = n()) %>%
              group_by(fr_bin) %>% 
              mutate(var_total_byfreq= sum(var_count_byfreq)) %>%
              ungroup() %>%
              mutate( var_mean_byfreq = var_count_byfreq / sum(var_count_byfreq),
                      total_mean = cumsum(var_mean_byfreq )) %>% 
              arrange(total_mean)

p0.0 = ggplot(snp_count, aes(x=factor(nvar))) +
     geom_bar(aes(y=var_count),stat='identity') + 
     xlab('# parallel variants')  + ylab('Count')
p0.1 = ggplot(snp_count, aes(x=factor(nvar))) +
     geom_bar(aes(y=var_mean),stat='identity') + 
     xlab('# parallel variants')  + ylab('variants %') +
     scale_y_continuous(labels = scales::percent) 

p1.0 = ggplot(snp_byfreq, aes(x=factor(nvar),group=fr_bin)) +
     geom_bar(aes(y=var_count_byfreq,fill=fr_bin),stat='identity') + 
     scale_fill_wsj(name='SNP%') + theme(legend.position = 'top', legend.text = element_text(size=9))+
     xlab('# parallel variants')  + ylab('Count')

p1.1 = p1 + facet_wrap(~fr_bin,scales = 'free_y') + theme(legend.pos='non')


p2.0 = ggplot(snp_byfreq, aes(x=fr_bin,fill=fr_bin)) + 
     geom_bar(aes(y=var_mean_byfreq),stat='identity')+
     scale_y_continuous(labels = scales::percent) +
     scale_fill_wsj(name='SNP%') + scale_color_wsj() + theme(legend.position = 'top', legend.text = element_text(size=9))+
     scale_x_discrete('Variant frequency bin')  + ylab('% variants')

p2.1 = ggplot(snp_byfreq, aes(x=1,fill=fr_bin)) + 
     geom_bar(aes(y=var_mean_byfreq),stat='identity', position = 'stack') +
     geom_text_repel(data=subset(snp_byfreq,nvar==1),segment.size = 1, direction = "y",
               mapping=aes(y=var_mean_byfreq,x=0,color=fr_bin,label=sprintf("%.f%% (%s)\n",100*var_mean_byfreq,var_total_byfreq)),
               position='stack',fontface='bold',show.legend = F)+
     scale_fill_wsj(name='SNP%') + scale_color_wsj() + theme(legend.position = 'top', legend.text = element_text(size=9))+
     scale_y_continuous(labels = scales::percent) +
     scale_x_discrete('Variant frequency bin')  + ylab('% variants')

# Histogram for number of SNP
p3.0 = get_snp_orf_plot(FREQMIN = 0)
p3.1 = get_snp_orf_plot()
# SNP count vs Protein Length
p4.0 = get_snp_protlen_plot(get_snp_protlen(PROT_SNP,0))
p4.1 = get_snp_protlen_anim(animated=T,psnp=PROT_SNP, width = 1920/2, height = 1080/2, units = "px")

# SNP count along protein length
p5.0=get_snp_pos_plot(FREQMIN = 0)
p5.1=get_snp_pos_plot(PROT_SNP)
```


Similarly, I calculated the number of "early-STOP" codons,
*i.e.* STOP codons which occur before the last amino acid.

```{r stop-plot}
stop_count = PROT_SNP %>% dplyr::filter(alt_fr>0.1) %>% 
        group_by(id) %>%
        summarize( nstop = sum(early_stop),len=mean(len)) %>%
        left_join(DESC,by=c('id'='ORF')) %>% 
        left_join(EXP,by=c('id'='prot')) %>%
        mutate(has_introns = (id %in% CAT.ORF$has_intron))
# Histogram for number of early STOP 
F6 = ggplot(stop_count,aes(x=nstop)) + 
      geom_histogram(binwidth = 1, fill=MAIN.COLOR,color='black') +
      xlab("early-STOP count") + ylab('# ORFS') + scale_y_log10()
plot(F6)

# early STOP count vs Protein Length
F7 = ggplot(stop_count %>% filter(nstop >0)) + 
     geom_point(aes(y=nstop,x=log10_mean.mpc,text=id,func=FUNCTION,color=factor(has_introns)), shape=21,size=0.8) + 
     xlab("Protein Length (AA)") + ylab('early-STOP count')
ggplotly(F7)
```

## SNP vs. evolution

```{r snp-evo}

R4S.desc = left_join(R4S,DESC,by=c('orf'='ORF')) %>%
           left_join(SGD,by=c('SGD'='sgdid')) %>%
           filter(complete.cases(r4s.yk11,r4s.fungi)) %>%
           ungroup() %>%
           mutate( has_introns=orf %in% CAT.ORF$has_intron,
                   has_ohnologs=orf %in% CAT.ORF$ohnologs,
                   sizerange= cut_number(len.s288c,n = 4),
                   is_dispensable = orf %in% CAT.ORF$dispensable,
                   is_essential = orf %in% CAT.ORF$essential
                  )# %>%
          #left_join(LOC.df,by=c('UNIPROT'='id')) %>%
          #add_count(loc,name = 'nloc') %>% filter(nloc > 100)
# CC=cor.sub.by(R4S.desc, XX='log10.r4s.yk11',YY='log10.r4s.fungi',BY='loc', ID = 'orf', na.rm = T) %>% 
#       mutate( toshow=sprintf("r=%.3f\np=%.1e\nn=%s",r,p,N) )

C6.0 =  spearman.toplot(Y=R4S.desc$log10.r4s.fungi,X=R4S.desc$log10.r4s.yk11)
p6.0 = ggplot(R4S.desc, aes(y=log10.r4s.fungi,x=log10.r4s.yk11,text=orf,func=FUNCTION)) + 
     geom_point(fill=MAIN.COLOR,alpha=0.4,shape=21,size=0.8) + 
     #geom_point(data=subset(R4S.desc,has_introns),color='red',fill=NA,shape=1,size=1.4,stroke=0.2)+
     #geom_point(data=subset(R4S.desc,has_ohnologs),color='blue',fill=NA,shape=1,size=1.4,stroke=0.2)+
     #geom_point(data=subset(R4S.desc,is_essential),color='green',fill=NA,shape=1,size=1.4,stroke=0.2)+
     #geom_point(data=subset(R4S.desc,is_dispensable),color='purple',fill=NA,shape=1,size=1.4,stroke=0.2)+
     geom_text(data=C6.0, aes(x=Inf,y=-Inf,label=toshow),color=MAIN.COLOR,size=4,inherit.aes = F,hjust='inward',vjust='inward') +
     #geom_text(data=subset(R4S.desc,outlier), mapping=aes(label=orf),color='white',size=2) +
     xlab("1011 strains evolutionary rate") + ylab('Fungi Evolutionary rate')  
#    facet_wrap(~loc)
plot(p6.0)
  

snp_count_orf=get_snp_orf(PROT_SNP,0)
SNP.PROT = left_join(snp_count_orf, EXP , by=c('id'='prot')) %>% # Add expression data
           left_join(R4S, by=c('id'='orf')) %>% # Add conservation data 
           left_join(DESC,by=c('id'='ORF')) # Add annotation data

# SNP count vs Evolutionary rate fungi
C6.1 =  spearman.toplot(SNP.PROT$sites,SNP.PROT$log10.r4s.fungi)
p6.1 = ggplot(SNP.PROT) + 
     geom_point(aes(x=sites,y=log10.r4s.fungi,text=id,func=FUNCTION), color=MAIN.COLOR, fill=NA,shape=21,size=0.8) + 
     geom_text(C6.1, mapping=aes(x=1,y=0.5,label=toshow),hjust='inward',vjust='inward',check_overlap =T,size=4) + 
     xlab('SNP count') + ylab("Fungi Evolutionary rate") +
     scale_x_continuous(trans='log2')
ggplotly(p6.1)
plot(p6.1)

# SNP count vs Evolutionary rate 1011 strains
C6.2 =  spearman.toplot(SNP.PROT$sites,SNP.PROT$log10.r4s.yk11)
p6.2= ggplot(SNP.PROT) + 
      geom_point(aes(x=sites,y=log10.r4s.yk11,text=id,func=FUNCTION), color=MAIN.COLOR, fill=NA,shape=21,size=0.8) + 
      geom_text(C6.2, mapping=aes(x=1,y=0.5,label=toshow),hjust='inward',vjust='inward',check_overlap =T,size=4) + 
      xlab('SNP count') + ylab("1011 strains evolutionary rate") + 
      scale_x_continuous(trans='log2')
ggplotly(p6.2)
plot(p6.2)
```

## SNP vs. expression

```{r snp-exp}
SNP.PROT$has_ortho = SNP.PROT$id %in% FUNGI$r4s_orf # filter for orthologs

# SNP count vs Median protein expression 
C7.0 =   cor.sub.by(SNP.PROT, "log10_median.mpc", "sites", 'has_ortho') %>%
         mutate( toshow = sprintf("r %.3f \n p %.1e \n n %s",r,p,n))
p7.0 = ggplot(SNP.PROT) + 
      geom_point(aes(y=sites,x=log10_median.mpc,text=id,func=FUNCTION), color=MAIN.COLOR, fill=NA,shape=21,size=0.8) + 
      geom_text(C7.0, mapping=aes(x=0,y=1,label=toshow),hjust='inward',vjust='inward',check_overlap =T,size=4) + 
      xlab("Molecules per cell (log10 median)") + ylab('SNP count') +
      scale_y_continuous(trans='log2') + facet_wrap(~has_ortho,labeller = label_both)
ggplotly(p7.0 )
plot(p7.0 )
# Evolutionary rate 1011 strains vs Median protein expression 
C7.1 =   cor.sub.by(SNP.PROT, "log10_median.mpc", "log10.r4s.yk11", 'has_ortho') %>%
        mutate( toshow = sprintf("r %.3f \n p %.1e \n n %s",r,p,n))

p7.1 = ggplot(SNP.PROT) + 
      geom_point(aes(y=log10.r4s.yk11,x=log10_median.mpc,text=id,func=FUNCTION), color=MAIN.COLOR, fill=NA,shape=21,size=0.8) +
      geom_text(C7.1, mapping=aes(x=0,y=-4,label=toshow),hjust='inward',vjust='inward',check_overlap =T,size=4) + 
      ylab("1011 strains evolutionary rate") + xlab('Molecules per cell (log10 median)') + facet_wrap(~has_ortho,labeller=label_both)
ggplotly(p7.1)
plot(p7.1)
# Evolutionary rate fungi vs Median protein expression 
C7.2 =  spearman.toplot(SNP.PROT$log10_median.mpc,SNP.PROT$log10.r4s.fungi)
p7.2 = ggplot(SNP.PROT) + 
      geom_point(aes(y=log10.r4s.fungi,x=log10_median.mpc,text=id,func=FUNCTION), color=MAIN.COLOR, fill=NA,shape=21,size=0.8) +
      geom_text(C7.2, mapping=aes(x=6,y=0.5,label=toshow),hjust='inward',vjust='inward',check_overlap =T,size=4) + 
     ylab("fungi evolutionary rate") + xlab('Molecules per cell (log10 median)') 
ggplotly(p7.2)
plot(p7.2)

```



## Variant viewer (Testing... in progress)
```{r snp-viewer, echo = FALSE}
library(shiny)
#shinyAppDir(
#  system.file("shiny/06_tabsets", package="shiny"),
#  options = list(width = "100%", height = 700)
#)
```

```{r all-plot, echo=F, include=F}
PATH.PLOT="plots/"
SAVED.PLOTS = list.files(PATH.PLOT)
SCALE_PLOT=1.2
EXT_GGPLOT2=c("eps", "ps", "tex" , "pdf", "jpeg", "tiff", "png", "bmp", "svg")

PLOTS = tribble( ~gplot, ~scalingf, ~fname, 
p0.0, SCALE_PLOT,     '0.0_var_count.png', 
p0.1, SCALE_PLOT,     '0.1_var_prop.png', 
p1.0, SCALE_PLOT,     '1.0_var_count_by_freq_stacked.png', 
p1.1, SCALE_PLOT,     '1.1_var_count_by_freq.png', 
p2.0, SCALE_PLOT,     '2.0_var_prop_by_freq_stacked.png', 
p2.1, SCALE_PLOT,     '2.1_var_prop_by_freq.png',  
p3.0, SCALE_PLOT,     '3.0_snp_orf.png',
p3.1, SCALE_PLOT,     '3.1_snp_orf_by_minfreq.png',
p4.0, SCALE_PLOT,     '4.0_snp_protlen.png', 
p4.1, SCALE_PLOT,     '4.1_snp_protlen.gif', 
p5.0, SCALE_PLOT,     '5.0_snp_pos.png', 
p5.1, SCALE_PLOT,     '5.1_snp_pos_by_minfreq.png', 
p6.0, SCALE_PLOT,     '6.0_evo_fungi_vs_pop.png', 
p6.1, SCALE_PLOT*0.6, '6.1_evo_fungi_snp.png', 
p6.2, SCALE_PLOT*0.8, '6.2_evo_pop_snp.png', 
p7.0, SCALE_PLOT,     '7.0_snp_exp.png', 
p7.1, SCALE_PLOT*0.8, '7.1_evo_pop_exp.png', 
p7.2, SCALE_PLOT*0.6, '7.2_evo_fungi_exp.png', 
)
RM.PLOTS = FALSE
if( interactive() ){ RM.PLOTS = askYesNo("About to delete all saved plots...", default=F) }
if( RM.PLOTS ){ unlink(file.path(PATH.PLOT,SAVED.PLOTS)) }
for( i in 1:nrow(PLOTS) ){
  S = PLOTS$scalingf[i]
  FILE_EXT = tolower(file_ext(PLOTS$fname[i]))
  if(FILE_EXT %in%  EXT_GGPLOT2){
    ggsave(path = PATH.PLOT, plot = PLOTS$gplot[[i]], filename = PLOTS$fname[i],scale =S )
  }else if(FILE_EXT == 'gif'){
    anim_save(path = PATH.PLOT, PLOTS$fname[i], animation =  PLOTS$gplot[[i]])
  }else{
    warning(sprintf("Unrecognized file extension for: %s",PLOTS$fname))
  }
}
```

# References {-}

Jackson et al.: 
[***Genome evolution across 1,011 Saccharomyces cerevisiae isolates***](https://doi.org/10.1038/s41586-018-0030-5)
<br>
Rate4site: 
[***Comparison of site-specific rate-inference methods: Bayesian methods are superior***](https://doi.org/10.1093/molbev/msh194)
<br>
Wapinski et al.: [***Natural history and evolutionary principles of gene duplication in fungi***](https://www.nature.com/articles/nature06107)
<br>
Ho et al.: [***Unification of Protein Abundance Datasets Yields a Quantitative Saccharomyces cerevisiae Proteome***](https://doi.org/10.1016/j.cels.2017.12.004)
<br>
Dos Reis et al.: [***Solving the riddle of codon usage preferences: a test for translational selection***](https://doi.org/10.1093/nar/gkh834)

# Session Info {-}

```{r session-info}
sessionInfo()
```
